File: Assets/Sequencer/Slot/KeyCube.js
#pragma strict

private var vibe = 0.0;
private var spin = 0.0;
private var initialScale = 1.0;

static private var baseColors = [
	Color(1.0, 1.0, 1.0),		// Pentatonic
	Color(0.7, 1.0, 0.7),		// Diatonic
	Color(0.7, 1.0, 1.0),		// Pentatonic + IIV
	Color(0.7, 0.7, 1.0),		// Major blues
	Color(1.0, 1.0, 0.7)		// Ryukyu
];

function Start() {
	initialScale = transform.localScale.x;
}

function SetColor(scaleIndex : int, degree : int) {
	var brightness = (degree & 1) ? 0.7 : 0.9;
	renderer.material.color = baseColors[scaleIndex % baseColors.Length] * brightness;
}

function Update() {
	transform.localRotation = Quaternion.AngleAxis(spin * 360.0, Vector3.up);
	transform.localScale = Vector3.one * (initialScale * (1.0 + 0.2 * vibe * Mathf.Sin(40.0 * Time.time)) * (1.0 - 0.5 * spin));
	vibe = ExpEase.Out(vibe, 0.0, -8.0);
	spin = ExpEase.Out(spin, 0.0, -8.0);
}

function KeyOn() {
	vibe = 1.0;
}

function RemoveArpies() {
	vibe = 1.0;
	spin = 1.0;
}

File: Assets/Sequencer/Slot/KeyCube.cs
using UnityEngine;

namespace Arpie {

class KeyCube : MonoBehaviour
{
    static Color [] BaseColors = new []
    {
        new Color(1, 1, 1),         // Pentatonic
        new Color(0.7f, 1, 0.7f),   // Diatonic
        new Color(0.7f, 1, 1),      // Pentatonic + IIV
        new Color(0.7f, 0.7f, 1),   // Major blues
        new Color(1, 1, 0.7f)       // Ryukyu
    };

    static int ColorKey = Shader.PropertyToID("_Color");

    Transform _mesh;
    Renderer _renderer;
    MaterialPropertyBlock _overrides;

    float _vibe;
    float _spin;

    void Awake()
    {
        _mesh = transform.Find("Mesh");
        _renderer = _mesh.GetComponent<Renderer>();
        _overrides = new MaterialPropertyBlock();
    }

    public void SetColor(int scaleIndex, int degree)
    {
        var value = (degree & 1) != 0 ? 0.7f : 0.9f;
        var color = BaseColors[scaleIndex % BaseColors.Length] * value;
        _overrides.SetColor(ColorKey, color);
        _renderer.SetPropertyBlock(_overrides);
    }

    void Update()
    {
        var scale = 1 + 0.2f * _vibe * Mathf.Sin(40 * Time.time);
        scale *= 1 - 0.5f * _spin;

        _mesh.localRotation = Quaternion.AngleAxis(_spin * 360, Vector3.up);
        _mesh.localScale = Vector3.one * scale;

        _vibe = ExpEase.Out(_vibe, 0, -8);
        _spin = ExpEase.Out(_spin, 0, -8);
    }

    void KeyOn()
      => _vibe = 1;

    void RemoveArpies()
      => (_vibe, _spin) = (1, 1);
}

} // namespace Arpie
